R/pkg_ref_class.R

```diff
@@ -68,35 +68,59 @@
 #' ref_2$source  # returns 'pkg_cran_remote'
 #' }
 pkg_ref <- function(x, ...) {
-  if (missing(x)) return(structure(logical(0L), class = "pkg_ref"))
+  if (missing(x)) {
+    empty <- new.env(parent = emptyenv())
+    class(empty) <- c("pkg_ref", class(empty))
+    return(empty)
+  }
+
   as_pkg_ref(x, ...)
 }
 
-
+#' @keywords internal
+pkg_ref_ptype <- function() {
+  ptype <- logical()
+  class(ptype) <- c("pkg_ref_ptype", "pkg_ref", class(ptype))
+  ptype
+}
```

Comment:
Instead of having `pkg_ref()` do double duty and using the empty call for our `ptype`, this just leaves that to a separate internal function. We give it a special `_ptype` class suffix to handle some internals that assume a `pkg_ref` is an environment().